src/ffi/web_socket/error.rs

```diff
@@ -0,0 +1,272 @@
+    /// received a close frame) on the server end and as soon as the server has closed the
+    /// underlying connection if this endpoint is a client.
+    ///
+    /// Thus when you receive this, it is safe to drop the underlying connection.
+    ///
+    /// Receiving this error means that the WebSocket object is not usable anymore and the
+    /// only meaningful action with it is dropping it.
+    #[error("Connection closed normally")]
+    ConnectionClosed,
+    /// Trying to work with already closed connection.
+    ///
+    /// Trying to read or write after receiving `ConnectionClosed` causes this.
+    ///
+    /// As opposed to `ConnectionClosed`, this indicates your code tries to operate on the
+    /// connection when it really shouldn't anymore, so this really indicates a programmer
+    /// error on your part.
+    #[error("Trying to work with closed connection")]
+    AlreadyClosed,
+    /// Input-output error. Apart from WouldBlock, these are generally errors with the
+    /// underlying connection and you should probably consider them fatal.
+    #[error("IO error: {io_error}")]
+    Io {
+        /// Input-output error. Apart from WouldBlock, these are generally errors with the
+        /// underlying connection and you should probably consider them fatal.
+        io_error: io::error::IoError,
+    },
+    /// TLS error.
+    #[error("TLS error: {tls_error}")]
+    Tls {
+        /// TLS error
+        tls_error: String,
+    },
+    /// - When reading: buffer capacity exhausted.
+    /// - When writing: your message is bigger than the configured max message size
+    ///   (64MB by default).
+    #[error("Space limit exceeded: {capacity_error}")]
+    Capacity {
+        /// When there are too many headers or the message is too long
+        capacity_error: CapacityError,
+    },
+    /// Protocol violation.
+    #[error("WebSocket protocol error: {protocol_error}")]
+    Protocol {
+        /// Protocol violation
+        protocol_error: ProtocolError,
+    },
+    /// Message send queue full.
+    #[error("Send queue is full")]
+    SendQueueFull {
+        /// The [Message] that could not fit in the send queue.  It should be resent later or the
+        /// error needs to propagate up.
+        message: Message,
+    },
+    /// UTF coding error.
+    #[error("UTF-8 encoding error")]
+    Utf8,
+    /// Invalid URL.
+    #[error("URL error: {url_error}")]
+    Url {
+        /// The URL error as a string
+        url_error: String,
+    },
+    /// HTTP error.
+    #[error("HTTP error: {}", response.status_code)]
+    Http {
+        /// Error response from the server.
+        response: http::Response,
+    },
+    /// HTTP format error.
+    #[error("HTTP format error: {error}")]
+    HttpFormat {
+        /// HTTP format error.
+        error: http::HttpError,
+    },
+}
+impl From<&tokio_tungstenite::tungstenite::Error> for WebSocketError {
```

Comment:
I think `use tokio_tungstenite::tungstenite::error::Error as TungsteniteError` and then use the aliased import is cleaner.